presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java

```diff
@@ -0,0 +1,419 @@
+                    List<TypeSignature> scalarAggregationSourceTypeSignatures = ImmutableList.of(
+                            symbolAllocator.getTypes().get(nonNullableAggregationSourceSymbol).getTypeSignature());
+                    functions.put(symbol, functionRegistry.resolveFunction(
+                            count,
+                            scalarAggregationSourceTypeSignatures,
+                            false));
+                }
+                else {
+                    aggregations.put(symbol, entry.getValue());
+                    functions.put(symbol, scalarAggregation.getFunctions().get(symbol));
+                }
+            }
+
+            List<Symbol> groupBySymbols = leftOuterJoin.getLeft().getOutputSymbols();
+            return Optional.of(new AggregationNode(
+                    idAllocator.getNextId(),
+                    leftOuterJoin,
+                    groupBySymbols,
+                    aggregations.build(),
+                    functions.build(),
+                    scalarAggregation.getMasks(),
+                    ImmutableList.of(groupBySymbols),
+                    scalarAggregation.getStep(),
+                    scalarAggregation.getSampleWeight(),
+                    scalarAggregation.getConfidence(),
+                    scalarAggregation.getHashSymbol()));
+        }
+
+        public Optional<RewrittenScalarAggregationSource> rewriteScalarAggregationSource(PlanNode aggregationSource, List<Symbol> correlation)
+        {
+            Optional<FilterNode> filter = searchFrom(aggregationSource)
+                    .where(FilterNode.class::isInstance)
+                    .skipOnlyWhen(isInstanceOfAny(ProjectNode.class, LimitNode.class))
+                    .findFirst();
+
+            if (!filter.isPresent()) {
+                return Optional.of(new RewrittenScalarAggregationSource(ImmutableList.of(), aggregationSource));
+            }
+
+            Expression predicate = filter.get().getPredicate();
+
+            if (!isSupportedPredicate(predicate)) {
+                return Optional.empty();
+            }
+
+            if (!DependencyExtractor.extractAll(predicate).containsAll(correlation)) {
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<Expression> newPredicate = ImmutableList.builder();
+            ImmutableList.Builder<Expression> joinExpressions = ImmutableList.builder();
+            for (Expression conjunct : extractConjuncts(predicate)) {
+                List<Symbol> symbols = DependencyExtractor.extractAll(conjunct);
+                if (correlation.stream().anyMatch(symbols::contains)) {
+                    joinExpressions.add(conjunct);
+                }
+                else {
+                    newPredicate.add(conjunct);
+                }
+            }
+
+            FilterNode newFilterNode = new FilterNode(
```

Comment:
Also, if you move this closer to where the filter node is used, it will make the code easier to read.